### PR TITLE
Proof that subst preserves evaluation to value

### PR DESCRIPTION
**Pull request by cdisselkoen**
A major lemma needed for the PE soundness proof; split into its own PR because it's large enough by itself.

### CODE REVIEW

File: cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean [hunk 1]
```diff
@@ -148,4 +149,158 @@ theorem on_concrete_eqv_concrete_eval {attrs : List (Attr × Spec.Expr)} {reques
...
+    have (khd, vhd) := hd ; clear hd
+    simp only at *
+    replace ⟨(khd', vhd'), h₁, h₂⟩ := h₁
+    cases vhd <;> simp only [Option.some.injEq, Prod.mk.injEq, exists_eq_left'] at *
+    replace ⟨h₁, h₁'⟩ := h₁ ; subst khd' vhd' ; rename_i vhd
+    replace ⟨tl', h₂, h₃⟩ := h₂
+    subst h₃
+    exists (tl'.map Prod.snd)
+    simp only [List.map_cons, and_true]
+    exact lemma h₂
```
Comment (emina) [inline cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean:271]: Nit: I was thrown off by the name `lemma` for this theorem, and read this line as a special form of the `exact` tactic (as in `exact lemma` vs just `exact`).

**Review by emina — APPROVED** [review]
LGTM.

File: cedar-lean/Cedar/Thm/Partial/Evaluation.lean [hunk 2]
```diff
@@ -141,3 +143,236 @@ theorem on_concrete_gives_concrete (expr : Spec.Expr) (request : Spec.Request) (
...
+    have ih : ∀ x ∈ xs, SubstPreservesEvaluationToConcrete x req req' entities subsmap := by
+      unfold SubstPreservesEvaluationToConcrete
+      intro x h₂ _ v' hx'
+      replace ⟨v, hx⟩ := hx x h₂
+      simp only [hx, Except.ok.injEq, Partial.Value.value.injEq] at hx'
+      subst v'
+      have := List.sizeOf_lt_of_mem h₂
+      exact subst_preserves_evaluation_to_value wf_r wf_e h_req hx
+    exact Call.subst_preserves_evaluation_to_value ih h_req v h₁
+termination_by expr
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-lean/Cedar/Thm/Partial/Evaluation.lean:378]: I think it's not needed.  But also in the event that the termination proof breaks, in my experience you get a much clearer error message this way than without `termination_by expr`.  Without it, Lean says something like "failed to find a termination measure", while with it, Lean shows the actual proof obligation that wasn't dispatched.  (Simulate the termination proof breaking by deleting one of the `have` lines for the `sizeOf` lemmas.)

File: cedar-lean/Cedar/Thm/Partial/Evaluation/Basic.lean [hunk 3]
```diff
@@ -85,11 +85,33 @@ end Cedar.Partial
...
+-/
+def SubstPreservesEvaluationToConcrete (expr : Partial.Expr) (req req' : Partial.Request) (entities : Partial.Entities) (subsmap : Map Unknown Partial.Value) : Prop :=
+  req.subst subsmap = some req' →
+  ∀ v, Partial.evaluate expr req entities = .ok (.value v) → Partial.evaluate (expr.subst subsmap) req' (entities.subst subsmap) = .ok (.value v)
+
+/--
+  Prop that a list of partial values is actually a list of concrete values
+-/
+def is_all_concrete (pvals : List Partial.Value) : Prop :=
+  ∃ vs, pvals.mapM (λ x => match x with | .value v => some v | .residual _ => none) = some vs
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/Basic.lean:115]: Is it definitionally equivalent?  Will it unify with existing expressions in the `match` form?

File: cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean (hunk 1 above)
Comment (cdisselkoen) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean:271]: Yeah, I should come up with a better name for this lemma

File: cedar-lean/Cedar/Thm/Partial/Evaluation/Ite.lean [hunk 4]
```diff
@@ -47,4 +49,73 @@ theorem on_concrete_eqv_concrete_eval {x₁ x₂ x₃ : Spec.Expr} {request : Sp
...
+            left
+            cases hx₂ : Partial.evaluate x₂ request entities
+            <;> simp only [hx₂, Except.ok.injEq] at h₁
+            case ok v₂ => subst h₁ ; simp
+          case false =>
+            right
+            cases hx₃ : Partial.evaluate x₃ request entities
+            <;> simp only [hx₃, Except.ok.injEq] at h₁
+            case ok v₃ => subst h₁ ; simp
+      case set s | record m => simp at h₁
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/Ite.lean]: Fixed

File: cedar-lean/Cedar/Thm/Partial/Evaluation/Or.lean [hunk 5]
```diff
@@ -54,4 +54,46 @@ theorem on_concrete_eqv_concrete_eval {x₁ x₂ : Spec.Expr} {request : Spec.Re
...
+    cases pval <;> simp only [Except.ok.injEq] at h₁
+    case value v =>
+      cases v
+      case prim p =>
+        cases p <;> simp only [Except.bind_ok, Except.bind_err] at h₁
+        case bool b =>
+          cases b
+          <;> simp only [reduceIte, Except.ok.injEq, Partial.Value.value.injEq] at h₁
+          <;> simp [h₁]
+      case set s | record m => simp at h₁
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/Or.lean]: Fixed

File: cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean (hunk 1 above)
Comment (cdisselkoen) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean:271]: Renamed. Let me know if you have a better suggestion

File: cedar-lean/Cedar/Thm/Partial/Evaluation.lean (hunk 2 above)
[the remark this answers is not in the document]
Comment (emina) [reply cedar-lean/Cedar/Thm/Partial/Evaluation.lean:378]: Yes, I usually include it while working on the proof and delete when done.  But there is no harm in keeping it. I was just curious if it was necessary.

**Review by emina — APPROVED** [review]

File: cedar-lean/Cedar/Thm/Partial/Evaluation/AndOr.lean [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+import Cedar.Partial.Evaluator
+import Cedar.Thm.Data.Control
+import Cedar.Thm.Partial.Evaluation.Basic
+
+namespace Cedar.Thm.Partial.Evaluation.AndOr
+
+open Cedar.Data
+open Cedar.Partial (Unknown)
+
+/- ## Lemmas shared by And.lean and Or.lean -/
```
Comment (khieta) [inline cedar-lean/Cedar/Thm/Partial/Evaluation/AndOr.lean:26]: Why do you need a new file here? Why can't you just define this in `And.lean` or `Or.lean` and then import the lemma in the other file, like how you do for Set & Call?

File: cedar-lean/Cedar/Thm/Partial/Evaluation/Basic.lean [hunk 7]
```diff
@@ -85,11 +85,33 @@ end Cedar.Partial
...
+  Prop that .subst preserves evaluation to a concrete value
+-/
+def SubstPreservesEvaluationToConcrete (expr : Partial.Expr) (req req' : Partial.Request) (entities : Partial.Entities) (subsmap : Map Unknown Partial.Value) : Prop :=
+  req.subst subsmap = some req' →
+  ∀ v, Partial.evaluate expr req entities = .ok (.value v) → Partial.evaluate (expr.subst subsmap) req' (entities.subst subsmap) = .ok (.value v)
+
+/--
+  Prop that a list of partial values is actually a list of concrete values
+-/
+def is_all_concrete (pvals : List Partial.Value) : Prop :=
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/Basic.lean]: Fixed

File: cedar-lean/Cedar/Thm/Partial/Evaluation/Ite.lean [hunk 8]
```diff
@@ -47,4 +49,73 @@ theorem on_concrete_eqv_concrete_eval {x₁ x₂ x₃ : Spec.Expr} {request : Sp
       cases p <;> simp only [Except.bind_ok, Except.bind_err]
       case bool b => cases b <;> simp
 
+/--
+  If partial-evaluating a `Partial.Expr.ite` produces `ok` with a concrete
+  value, then so would partial-evaluating any of the operands, unless the `ite`
+  short-circuits
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/Ite.lean]: Adjusted this doc comment

File: cedar-lean/Cedar/Thm/Partial/Evaluation/Var.lean [hunk 9]
```diff
@@ -84,4 +87,131 @@ theorem on_concrete_eqv_concrete_eval (v : Var) (request : Spec.Request) (entiti
...
+
+/--
+  If partial-evaluation of a `Var` returns a concrete value, then it returns the
+  same value after any substitution of unknowns
+-/
+theorem subst_preserves_evaluation_to_value {var : Var} {req req' : Partial.Request} {entities : Partial.Entities} {v : Spec.Value} {subsmap : Map Unknown Partial.Value}
+  (wf : req.AllWellFormed) :
+  req.subst subsmap = some req' →
+  Partial.evaluate (Partial.Expr.var var) req entities = .ok (.value v) →
+  Partial.evaluate ((Partial.Expr.var var).subst substmap) req' (entities.subst subsmap) = .ok (.value v)
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/Var.lean]: Good call, refactored

File: cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean [hunk 10]
```diff
@@ -148,4 +149,159 @@ theorem on_concrete_eqv_concrete_eval {attrs : List (Attr × Spec.Expr)} {reques
...
+      exists v
+
+/--
+  Lemma:
+
+  Inductive argument that if `mapM` on a list of attrs produces `.ok` with a
+  list of concrete vals, then it produces the same list of concrete vals after
+  any substitution of unknowns
+-/
+theorem mapM_subst_preserves_evaluation_to_values {attrs : List (Attr × Partial.Expr)} {req req' : Partial.Request} {entities : Partial.Entities} {subsmap : Map Unknown Partial.Value}
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean:190]: It's hard to generalize, for the same reasons that `map_attach2` and `map_attach2_snd` can't easily be interchanged.  `Partial.evaluate` doesn't apply to a (attr, expr) pair, so the lambda has a different form; the `IsAllConcrete` predicate has to be phrased to be over just the second component of each pair; the `subst` operation in the last line has to apply to only the second component of each pair

[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean:190]: Perhaps we could define `IsAllConcrete` and `subst` and `Partial.evaluate` and etc for `(Attr, Partial.Expr)` pairs.  Not sure if that would or would not be enough to generalize.  What would be the type of the first implicit arg to this theorem.

File: cedar-lean/Cedar/Thm/Partial/Evaluation/AndOr.lean (hunk 6 above)
Comment (khieta) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/AndOr.lean:26]: I guess I slightly prefer reusing existing files over making a new file. But I don't feel strongly. Just wanted to know if there was some reason I was missing 🙂

File: cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean (hunk 10 above)
[the remark this answers is not in the document]
Comment (khieta) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean:190]: Ack. If it's too much work to unify, then it's not a big deal. Maybe change the name though (e.g., `*_snd`) so people don't get confused if grepping through the repo to find the definition. (Yes, I know I should use VSCode's "go to definition" functionality, but sometimes I want to see everywhere a lemma is used!)

**Review by khieta — APPROVED** [review]

File: cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean (hunk 10 above)
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean]: Renamed

File: cedar-lean/Cedar/Thm/Partial/Evaluation/AndOr.lean (hunk 6 above)
Comment (cdisselkoen) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/AndOr.lean:26]: My instinct is generally to prefer many smaller files over fewer long files, but I don't feel too strongly either

Comment (cdisselkoen) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/AndOr.lean:26]: I think I'll leave it this way for this PR, and in a followup we can standardize on either one or the other approach

**Review by khieta — APPROVED** [review]

**Pull request merged**